### PR TITLE
Rewrite the extension code in Cython

### PR DESCRIPTION
**Pull request by luizirber**
Might work on #88 here too, moving methods from `Estimators` to `_minhash.MinHash` (even copying the Python code is OK now that it's all Cython).

Also #87? Using properties to expose data from C++ (see https://github.com/dib-lab/sourmash/compare/refactor/cython?expand=1#diff-1614fcd5af51f570309a55db3b62a123R58 for an example with `seed`)
 
## TODO

- [ ] Find a way to detect if Cython is installed, and if it is not use the precompiled `.cpp` file
- [x] Fix tests under 2.7
- [x] should we allow passing `bytes` to `add_sequence` and other methods accepting strings?

##
- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

File: sourmash_lib/kmer_min_hash.hh [hunk 1]
```diff
@@ -9,9 +9,17 @@
 #include <string>
 #include <unordered_map>
 
-// #include "kmer_hash.hh"
+#include "../third-party/smhasher/MurmurHash3.h"
 
-////
+uint64_t _hash_murmur(const std::string& kmer,
```
Comment (luizirber) [inline sourmash_lib/kmer_min_hash.hh]: This was defined in `sourmash_lib/_minhash.cc` before, so it wasn't visible when using a pure C++ program (only worked when header was compiled together with the extension...)

**Comment by luizirber** [comment]
Cheated to fix tests under Py27: Added `unicode_literals` where needed, but this won't work for anyone using the module without doing the same.

Now there is only one test failing, because `doctest` doesn't respect `unicode_literals`. Possible solution on the end of this topic: http://python3porting.com/problems.html#running-doctests

(But we need to add tests checking `bytes` passed to `add_sequence` and similar...)

**Comment by luizirber** [comment]
No need to fix the doctest, I refactor the methods to accept `str`, `unicode` and `bytes` on both Py3 and Py2.

**Comment by luizirber** [comment]
On second thought, I propose we merge this and work on #87 and #88 later.

Ready for review and merge @ctb @lgautier

**Comment by lgautier** [comment]
@luizirber a quick question was a benchmark ran to check the impact on speed and memory usage ?

**Comment by lgautier** [comment]
Hi again, I am assuming that the idea here is to allow easier experimentation. Is moving everything to Cython necessary (and helpful) ? Wouldn't keeping C or C++-written primitives make sense ? 

I am seeing that there would be an advantage to having mid or higher-level structures easy to modify and extend (and allow pure Python with options to more to Cython), but the primitives are what may need attention to optimization and my experience with Cython has been that all it is starting easy, but it is getting in the way later if ones wants to go into advanced optimization (it was few years ago though).

**Comment by betatim** [comment]
There are two parts: the bridge between python and the compiled extension, and the extension code itself.

I think the biggest gains are in using cython for the bridge, and there is no harm in using it for the code doing the heavy lifting. When the day comes to make things faster it means moving the classes used by the bridge to c++ but keeping the interface as cython. I would boldly claim that "you re doing it wrong" if cython for the bridge turns out to be the bottleneck 😀

**Comment by luizirber** [comment]
@lgautier: like @betatim said, the goal is to replace the bridge, not the primitives (which will continue to be C++)

That said, I'm writing some benchmarks using [asv][1], I'll make a PR soon.

[1]: https://github.com/spacetelescope/asv

**Comment by lgautier** [comment]
Got it. I had understood the title "Rewrite the extension code in Cython" as an intent to move everything to Cython.
(I did not not about asv. Looks interesting/useful. Thanks).

**Comment by luizirber** [comment]
Preliminary results using the new asv benchmarks on PR #100:

```
$ asv compare 2416946b10c84c4ed53e4efa03fc7841806de9ce 2fa30b234a526d00e21c77669e30bebad18bd6c4

All benchmarks:

    before     after       ratio
  [2416946b] [2fa30b23]
     37.81w     37.26w      0.99  benchmarks.PeakmemMinAbundanceSuite.peakmem_add_sequence
     37.94w     37.16w      0.98  benchmarks.PeakmemMinHashSuite.peakmem_add_sequence
-    2.65ms     1.70ms      0.64  benchmarks.TimeMinAbundanceSuite.time_add_hash
    15.47ms    15.06ms      0.97  benchmarks.TimeMinAbundanceSuite.time_add_sequence
+   26.94ms    29.92ms      1.11  benchmarks.TimeMinAbundanceSuite.time_compare
    15.24ms    15.33ms      1.01  benchmarks.TimeMinAbundanceSuite.time_concat
     failed     failed       n/a  benchmarks.TimeMinAbundanceSuite.time_copy
    27.23ms    25.93ms      0.95  benchmarks.TimeMinAbundanceSuite.time_count_common
+   11.68ms    13.43ms      1.15  benchmarks.TimeMinAbundanceSuite.time_get_mins
-   46.59ms    39.54ms      0.85  benchmarks.TimeMinAbundanceSuite.time_get_mins_abundance
    15.10ms    15.14ms      1.00  benchmarks.TimeMinAbundanceSuite.time_merge
     failed     failed       n/a  benchmarks.TimeMinAbundanceSuite.time_set_abundances
-    4.02ms     2.93ms      0.73  benchmarks.TimeMinHashSuite.time_add_hash
    18.45ms    17.47ms      0.95  benchmarks.TimeMinHashSuite.time_add_sequence
    28.30ms    29.09ms      1.03  benchmarks.TimeMinHashSuite.time_compare
-   15.76ms    13.46ms      0.85  benchmarks.TimeMinHashSuite.time_concat
     failed     failed       n/a  benchmarks.TimeMinHashSuite.time_copy
    30.16ms    27.52ms      0.91  benchmarks.TimeMinHashSuite.time_count_common
+   11.59ms    12.75ms      1.10  benchmarks.TimeMinHashSuite.time_get_mins
    14.76ms    13.80ms      0.94  benchmarks.TimeMinHashSuite.time_merge
```

I'll publish the HTML report soon, but most times are still similar before and after Cython.

**Comment by lgautier** [comment]
@luizirber : thanks. I was genuinely asking as a way to point out that gain in running speed can be measured ideally (and it helps with deciding to move forward with a change). I have played a little with design ideas and I have observed that just moving to C (or Cython, since the performances are similar) might not be the most important factor (I get pretty much the same of Python). Because of that, I'd say pick Cython if it makes it convenient for the development rather than for performance reasons.

**Pull request merged**